src/utils/ref_counted.h

```diff
@@ -0,0 +1,65 @@
+//-----------------------------------------------------------------------------
+// Copyright 2022 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//-----------------------------------------------------------------------------
+#ifndef SRC_UTILS_REF_COUNTED_H_
+#define SRC_UTILS_REF_COUNTED_H_
+
+#include <iostream>
+
+#include "src/utils/ref_counter.h"
+
+namespace raksha {
+
+// Specialize this class for your type if you want custom reference counting.
+template <typename T>
+class RefCountManager {
+  static void Retain(T* ptr) { ptr->Retain(); }
+  static void Release(T* ptr) { ptr->Release(); }
+};
+
+// A helper class that allows a type `T` to be reference counted for an
+// `intrusive_ptr`.
+template <typename T>
+class RefCounted {
+ protected:
+  RefCounted() : count_(0) {}
+  RefCounted(const RefCounted& other) : RefCounted() {}
+  RefCounted(const RefCounted&& other) : RefCounted() {}
+
+  RefCounted& operator=(const RefCounted&) = delete;
+
+ private:
+  void Retain() { count_.FetchAdd(1); }
+
+  void Release() {
+    CHECK(count_ != 0) << "Reference count is already zero.";
+    size_t old = count_.FetchSub(1);
+    if (old == 1) {
+      delete static_cast<T*>(this);
```

Comment:
I don't understand how this is supposed to work. In this context, `this` is the `RefCounted` class, which contains within it only a `RefCounter` field. If `T` has a non-trivial dtor, the field alignments would be all off.

The usual design I see for refcounted objects is that the ref counting header (which I assume this is) assumes that it is allocated directly before an instance of its payload (T in this context). So I assume that the fields should be more like:

```
mutable RefCounter count_;
T payload_;
```

and the `Release()` method would simply do:

```
delete this;
```

Reply:
Classes that need to reference count are supposed to inherit from this class (or define their own reference counting mechanism with `RefCountManager`). See https://github.com/google-research/raksha/pull/286/files#diff-50fe78aea3d9579425f207389cd8a115614f50882838f6fed0aa7e83f496023d

For this to work correctly, the destructor needed to be virtual.  I have fixed it now. The type `T` will be provided as part of the template. In any case, I removed the `static_cast` to avoid confusion. Given that destructor is virtual, I am not sure if `static_cast` helps at all.